### PR TITLE
Security: Harden websocket 'CheckOrigin'

### PR DESCRIPTION
**Pull request by mateea326**
## 📝 Description

This PR hardens the security of the Pico Protocol WebSocket channel by changing the default CheckOrigin behavior. Previously, the server would allow all origins by default if AllowOrigins was empty, creating a potential Cross-Site WebSocket Hijacking (CSWSH) vulnerability.

With this change:

If AllowOrigins is empty (default), the server now only accepts connections from the same origin (matching the Host header).
Users who intentionally want to allow all origins must now explicitly set AllowOrigins: ["*"].
This aligns the implementation with security best practices and gorilla/websocket's recommended defaults.

## 🗣️ Type of Change
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 📖 Documentation update
- [ ] ⚡ Code refactoring (no functional changes, no api changes)

## 🤖 AI Code Generation
- [ ] 🤖 Fully AI-generated (100% AI, 0% Human)
- [ ] 🛠️ Mostly AI-generated (AI draft, Human verified/modified)
- [x] 👨‍💻 Mostly Human-written (Human lead, AI assisted or none)

## 🔗 Related Issue

N/A

## 📚 Technical Context (Skip for Docs)
- **Reference URL:** https://owasp.org/www-project-web-security-testing-guide/v42/4-Web_Application_Security_Testing/11-Client-side_Testing/10-Testing_WebSockets.html
- **Reasoning:** Local tools running on localhost are still vulnerable to CSWSH if a user visits a malicious website while the tool is running. Enforcing Same-Origin by default ensures that only trusted local dashboards can talk to the WebSocket.

## 🧪 Test Environment
- **Hardware:**  N/A
- **OS:**  N/A
- **Model/Provider:**  N/A
- **Channels:**  pico

## 📸 Evidence (Optional)
<details>
<summary>Click to view Logs/Screenshots</summary>

```
=== RUN   TestOriginCheckVulnerability
=== RUN   TestOriginCheckVulnerability/Default_(empty_AllowOrigins)_should_REJECT_different_origin_(The_Fix)
09:36:00 ERR pico pico.go:358 > WebSocket upgrade failed error="websocket: request origin not allowed by Upgrader.CheckOrigin"
Connection from malicious-site.com correctly rejected by default (Fix verified!)
=== RUN   TestOriginCheckVulnerability/Default_(empty_AllowOrigins)_should_ALLOW_same-origin
Connection from same-origin succeeded as expected
=== RUN   TestOriginCheckVulnerability/Allow_all_when_AllowOrigins_is_explicitly_set_to_'*'
Explicit '*' correctly allows connection from any-site.com
=== RUN   TestOriginCheckVulnerability/Respect_AllowOrigins_when_configured
Connection from malicious-site.com correctly rejected with 403 Forbidden
Connection from trusted-site.com succeeded as expected
--- PASS: TestOriginCheckVulnerability (0.00s)
```

</details>

## ☑️ Checklist
- [x] My code/docs follow the style of this project.
- [x] I have performed a self-review of my own changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by mateea326 — COMMENTED** [review]
deleted semgrep logs

File: pkg/channels/pico/pico.go [hunk 1]
```diff
@@ -73,10 +74,19 @@ func NewPicoChannel(cfg config.PicoConfig, messageBus *bus.MessageBus) (*PicoCha
...
-			return true // allow all if not configured
+			if origin == "" {
+				return true
+			}
+			u, err := url.Parse(origin)
+			if err != nil {
+				return false
+			}
+			return u.Host == r.Host
 		}
```
[the remark this answers is not in the document]
Comment (mateea326) [reply pkg/channels/pico/pico.go]: @copilot apply changes based on [this feedback](https://github.com/sipeed/picoclaw/pull/2256#discussion_r3023180350)

**Comment by mateea326** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/sipeed/picoclaw/pull/2256#pullrequestreview-4046433382)

**Review by yinwm — CHANGES_REQUESTED** [review]
Thanks for the security hardening — the direction is correct and the change is well-scoped.

Before merging, there are two issues to address:

### 1. Hostname comparison is case-sensitive (pico.go:87)

`u.Host == r.Host` is a case-sensitive comparison, but RFC 3986 specifies that hostnames are case-insensitive. A request with `Origin: https://EXAMPLE.COM` and `Host: example.com` would be rejected.

**Suggested fix (pick one):**
- Use `strings.EqualFold(u.Host, r.Host)` for case-insensitive comparison
- Or, even simpler: when `AllowOrigins` is empty, leave `CheckOrigin` unset (`nil`). gorilla/websocket's built-in default already handles same-origin checking correctly, including case normalization and port handling. You'd only need the custom function when explicit `AllowOrigins` are configured.

### 2. Outdated comment in web/backend/api/pico_test.go:94-95

The comment "CheckOrigin allows all when the list is empty, so the channel still works" is no longer accurate after this PR. Please update it to reflect the new same-origin-by-default behavior.

The second option above (leaving `CheckOrigin` nil for the default case) would also simplify the tests, since you wouldn't need to assert `CheckOrigin != nil` for the empty-origins case.

Once addressed, this is good to merge.

File: pkg/channels/pico/pico_test.go [hunk 2]
```diff
@@ -240,3 +241,39 @@ func (c *PicoChannel) addConnForTest(pc *picoConn) {
...
+
+func TestNewPicoChannel_SetsCheckOriginWithWildcard(t *testing.T) {
+	// (2) explicit AllowOrigins=["*"] allows any origin
+	cfg := config.PicoConfig{AllowOrigins: []string{"*"}}
+	cfg.SetToken("test-token")
+
+	ch, err := NewPicoChannel(cfg, bus.NewMessageBus())
+	if err != nil {
+		t.Fatalf("NewPicoChannel: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (mateea326) [reply pkg/channels/pico/pico_test.go:264]: @copilot apply changes based on [this feedback](https://github.com/sipeed/picoclaw/pull/2256#discussion_r3123759095)

**Comment by mateea326** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/sipeed/picoclaw/pull/2256#pullrequestreview-4046460472)

**Comment by mateea326** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/sipeed/picoclaw/pull/2256#pullrequestreview-4155722213)
